[sidecar] Prevent coordinator crashing in non-sidecar enabled clusters by making sidecar plugin functionalities config driven

pdabre12 opened the pull request:
## Description
Prevent coordinator crashing in non-sidecar enabled clusters by making sidecar plugin functionalities config driven

## Motivation and Context
Solves part of #24964 

## Impact
Users would be able to include the ` presto-native-sidecar-plugin` plugin in the build even for non-sidecar enabled clusters.

## Test Plan
Unit tests

## Contributor checklist

- [x] Please make sure your submission complies with our [contributing guide](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md), in particular [code style](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#code-style) and [commit standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes
```
== NO RELEASE NOTE ==
```

pdabre12 commented:
Logs from running PrestoServer locally:

```
2025-04-23T15:51:11.214-0700	INFO	main	com.facebook.presto.server.PluginManager	Registering connector hudi
2025-04-23T15:51:11.216-0700	INFO	main	com.facebook.presto.server.PluginManager	-- Finished loading plugin ../presto-hudi/pom.xml --
2025-04-23T15:51:11.216-0700	INFO	main	com.facebook.presto.server.PluginManager	-- Skipping loading disabled plugin ../presto-native-sidecar-plugin/pom.xml --
2025-04-23T15:51:11.227-0700	INFO	main	com.facebook.presto.metadata.StaticCatalogStore	-- Loading catalog properties etc/catalog/blackhole.properties --
```

rschlussel commented:
The general model for all plugins has been that you should be able to load plugins without having any side effects if they aren't enabled. Also, as far as cluster configuration, it's inconvenient to have to disable the plugin in addition to setting use side car to false.  instead, i think it would be better for the NativeSessionpropertyProviders to have different names (currently the coordinator based one and the sidecar based one are both called native-worker), and loadSystemSessionProperty providers only loads the providers that are relevant for the configuration (similar to how AccessControlManager.loadSystemAccessControl(), but with a different/slightly more complicated combination of configs.

pdabre12 commented:
@rschlussel Thanks for the response.
I think it might be much simpler to have annotations on the sidecar specific functionalities as shown in this PR.
We can mark the other functionalities aka. NativeTypeManager, NativeFunctionNamespaceManager as dependent on `coordinator-sidecar-enabled` as well.
From a user's POV, all they would need to do even if the plugin was loaded is set `coordinator-sidecar-enabled` to false, to not use the sidecar functionalities. What do you think?

rschlussel reviewed (approved):
This seems like a fine option

1. is it possible to add a test that everything starts up properly now if the plugin is loaded but the config is not enabled?  
2. For some sidecar features it might be nice to have an additional flag that enables them (e.g. you want to disable sidecar constant folding at first, but still use the function registry).  Would this kind of approach still work if something needs two configs to be enabled?

pdabre12 commented:
@rschlussel 
Added the unit test.
The logic is to install the `NativeSidecarPlugin` on both the java and c++ clusters but not set the coordinator-sidecar-enabled(default false). If we are still able to run queries , means both the clusters were up without any issues. Will this work?

> For some sidecar features it might be nice to have an additional flag that enables them (e.g. you want to disable sidecar constant folding at first, but still use the function registry). Would this kind of approach still work if something needs two configs to be enabled?

For supporting multiple configs, added a new annotation which handles that, still could not find out how to unit test this though. Maybe we can let this change pass, create an issue and handle the unit test at a later time?

pdabre12 commented:
We only need the annotation for `NativeTypeManager` and the `NativeSessionPropertyProvider`, the other functionalities aka `NativeFunctionNamespaceManager` (using the StaticFunctionNamespaceStore) and `NativePlanChecker` are loaded through a configuration file and if these configuration files aren't present, we will not load them.

ZacBlanco commented on presto-common/src/main/java/com/facebook/presto/common/ConditionalOnConfigs.java:
```diff
@@ -0,0 +1,26 @@
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+public @interface ConditionalOnConfigs
```
Why do we need this variant?

pdabre12 replied:
It handles a list of `@ConditionalOnConfig`.
The idea is to have the ability to have additional config flags on the sidecar functionalities to conditionally load individual functionalities.
As in let's say on the `NativeFunctionNamespaceManager`, we have two `@ConditionalOnConfig` configs `coordinator-sidecar-enabled` and `load-function-registry`.

ZacBlanco commented on presto-main-base/src/main/java/com/facebook/presto/metadata/FunctionAndTypeManager.java:
```diff
@@ -420,8 +427,11 @@ public void loadTypeManager(String typeManagerName)
     {
-        for (String typeManagerName : typeManagerFactories.keySet()) {
-            loadTypeManager(typeManagerName);
+        Map<String, String> configMap = ImmutableMap.of("coordinator-sidecar-enabled", String.valueOf(serverConfig.isCoordinatorSidecarEnabled()));
```
Do we have a static constant anywhere that we can reference rather than repeating this string?

pdabre12 replied:
No, but can make one.

ZacBlanco commented on presto-main-base/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java:
```diff
@@ -150,8 +157,13 @@ public void loadSessionPropertyProvider(String sessionPropertyProviderName, Opti
     {
-        for (String sessionPropertyProviderName : workerSessionPropertyProviderFactories.keySet()) {
-            loadSessionPropertyProvider(sessionPropertyProviderName, functionAndTypeManager, nodeManager);
+        Map<String, String> configMap = ImmutableMap.of("coordinator-sidecar-enabled", String.valueOf(serverConfig.isCoordinatorSidecarEnabled()));
```
I'm wondering if there's a better way to load the map of properties so that this code is more generic. It seems tailored specifically for the sidecar. If any other plugin needed to make use of ConditionalOnConfig, the property would need to get added here which is not ideal.

pdabre12 replied:
It seems like the best option.
We could have the factories provide the required properties, but we would need to push the config values down to the factories, which doesn't seem to be possible.

pdabre12 replied:
It might be possible to have just one generic configMap with the properties, because if the properties aren't present in the provider class we just assume it to be true

pdabre12 replied:
@ZacBlanco 
What if we get a list of all the config properties using `injector.getInstance(ConfigurationFactory.class).getProperties()`; in `PrestoServer.java` and pass it down to `PluginManager.java`.
While loading the plugin, check the annotations to see if all the properties are satisfied, if yes load it.

Alternatively , if we don't want to use annotations, we can add a `getRequiredConfigs()` method in both `Plugin.java `and `CoordinatorPlugin.java` which would return a map of required properties to load this plugin.
Any plugin can override this method, returns its own set of configs, which can be validated before loading the plugin, making this validation generic for all plugins.
What do you think of this approach?

pdabre12 replied:
However, for the second approach, loading individual components of plugins like NativeSidecarPlugin might be tricky. For that use case, the annotations approach would be easier to handle.

rschlussel replied:
I think the second approach should be okay.  If you have the sidecar configured, as long as it's possible to *set* the particular component (e.g. constant folding thing) to the non-sidecar version without issue, it should be okay to have it loaded as long as there's no side effect. I think just making sure different implementations of the component (e.g. java builtin vs. sidecar) are registered different names should generally be enough.

pdabre12 replied:
How to control the loading of functionalities  of the sidecar? Like loading the function registry without the constant folding thing.
For eg: If you do not want to use `NativeTypeManager` and you load the `NativeSidecarPlugin` with `coordinator-sidecar-enabled` true, we load the `NativeTypeManager` which is the default type manager now, in this case, we want to skip loading the native type manager itself.

rschlussel replied:
i think it's okay if it's loaded as long as it's not configured as the one to use and loading it in itself won't break anything.

rschlussel replied:
and yes getRequiredConfigs() approach.

pdabre12 replied:
With the way the current code is, we can only have one between the `NativeWorkerSessionPropertyProvider` or pull the session properties from the sidecar. If we enable the sidecar, ideally we should not have the `NativeWorkerSessionPropertyProvider` loaded up. In the case we do, there will be a duplication of the session properties even if the session property providers are under different names. 
Maybe we can choose a mix of both the approaches, have `@ConditionalOnConfig` annotations for the inidvidual functionalities and have `getRequiredConfigs`() at the Plugin  level.

However, I suggest we let this one in , this change should skip the loading of the plugin if the necessary configs aren't satisfied. We can add the support for loading the individual functionalities of a plugin as a follow-up, depending on some more discussion.

amitkdutta commented:
@pdabre12 @ZacBlanco @rschlussel Folks will be really helpful if we can merge this one. Without this change, we can't include the plugin in build and can't test out any of the good changes coming with it. CC: @kewang1024

pdabre12 commented on presto-native-sidecar-plugin/src/test/java/com/facebook/presto/sidecar/TestNativeSidecarPlugin.java:
```diff
@@ -69,7 +69,11 @@ protected QueryRunner createQueryRunner()
             throws Exception
     {
-        return PrestoNativeQueryRunnerUtils.createJavaQueryRunner();
+        DistributedQueryRunner queryRunner = (DistributedQueryRunner) PrestoNativeQueryRunnerUtils.createJavaQueryRunner();
```
@rschlussel This should test the case where the plugin is included but is not loaded because the configs weren't satisfied. What do you think?

rschlussel replied:
the test seems okay, but would be nice to make it more explicit (single test that tests it).

pdabre12 replied:
Added an explicit test.

ZacBlanco commented on presto-spi/src/main/java/com/facebook/presto/spi/Plugin.java:
```diff
@@ -153,4 +155,9 @@ default Iterable<ClientRequestFilterFactory> getClientRequestFilterFactories()
         return emptyList();
     }
+
+    default Map<String, String> getRequiredConfigs()
```
Can we add javadoc for this method?

ZacBlanco commented on presto-spi/src/main/java/com/facebook/presto/spi/CoordinatorPlugin.java:
```diff
@@ -52,4 +55,9 @@ default Iterable<TypeManagerFactory> getTypeManagerFactories()
         return emptyList();
     }
+
+    default Map<String, String> getRequiredConfigs()
```
Can we add javadoc for this?

ZacBlanco commented on presto-native-sidecar-plugin/src/main/java/com/facebook/presto/sidecar/NativeSidecarPlugin.java:
```diff
@@ -51,6 +54,12 @@ public Iterable<FunctionNamespaceManagerFactory> getFunctionNamespaceManagerFact
+    @Override
+    public Map<String, String> getRequiredConfigs()
+    {
+        return ImmutableMap.of("coordinator-sidecar-enabled", "true");
```
Let's make `"coordinator-sidecar-enabled"` a constant somewhere?

ZacBlanco commented on presto-main-base/src/main/java/com/facebook/presto/server/PluginManager.java:
```diff
@@ -384,6 +394,11 @@ public void installPlugin(Plugin plugin)
     public void installCoordinatorPlugin(CoordinatorPlugin plugin)
     {
+        if (!loadPluginIfConfigured(plugin.getRequiredConfigs())) {
+            log.info("Skipped loading %s because some conditions weren't satisfied", plugin.getClass().getName());
```
Maybe let's log inside the `loadPluginIfConfigured` instead? Then we could log the configuration property and value which was not satisfied

ZacBlanco commented on presto-main-base/src/main/java/com/facebook/presto/server/PluginManager.java:
```diff
@@ -267,6 +272,11 @@ else if (plugin instanceof CoordinatorPlugin) {
     public void installPlugin(Plugin plugin)
     {
+        if (!loadPluginIfConfigured(plugin.getRequiredConfigs())) {
+            log.info("Skipped loading %s because some conditions weren't satisfied", plugin.getClass().getName());
```
Can remove if we put this inside `loadPluginIfConfigured`

ZacBlanco reviewed (changes requested):
Thanks for this improvement @pdabre12 . I like this approach. Just a few comments

rschlussel commented on presto-main-base/src/main/java/com/facebook/presto/server/PluginManager.java:
```diff
@@ -506,4 +521,20 @@ private void processPlugins(Artifact artifact, ClassLoader classLoader, String s
         }
     }
+
+    private boolean loadPluginIfConfigured(Map<String, String> expectedProperties)
```
I found the name confusing since this doesn't do the loading. Can we rename it? Maybe something like `pluginIsEnabled`.

rschlussel reviewed (commented):
the approach looks good. Thanks!

pdabre12 commented:
@rschlussel @ZacBlanco 
Thanks for the review.
Addressed your comments, please have another look.

aditi-pandit commented on presto-native-sidecar-plugin/src/test/java/com/facebook/presto/sidecar/TestNativeSidecarPluginWithCoordinatorSidecarDisabled.java:
```diff
@@ -0,0 +1,75 @@
+
+    @Test
+    public void testBasicQueries()
+    {
```
Would any of these queries be affected if the side-car plugin functionality was loaded ? It would be good to have a test which works when the side-car is enabled but does not otherwise.

pdabre12 replied:
Added a test case.

aditi-pandit reviewed (commented):
Thanks @pdabre12 for this code. Have a quick question about testing.

aditi-pandit reviewed (dismissed):
Thanks @pdabre12

pdabre12 commented:
@ZacBlanco @rschlussel Can one of you please approve again?

tdcmeehan commented on presto-spi/src/main/java/com/facebook/presto/spi/Plugin.java:
```diff
@@ -153,4 +155,15 @@ default Iterable<ClientRequestFilterFactory> getClientRequestFilterFactories()
+     *
+     * @return a map of required configuration keys and expected values for the plugin.
+     **/
+    default Map<String, String> getRequiredConfigs()
```
Right now, this is just used to binary enable/disable the plugin.  Instead of making it freeform string, and then doing a deep comparison to ensure both keys and values match, why not introduce a new type, say, `PluginConfiguration`, that has just one field, `enabled`?

As a Plugin author, this method seems confusing, because it's only used to enable or disable the plugin, yet the contract appears much wider ranging, accepting an arbitrary string to string Map.

pdabre12 replied:
I changed the approach altogether.

pdabre12 replied:
Instead of asking the Plugin author to provide the required set of configs, we would use config files to load individual functionalities of the sidecar. What do you think?

amitkdutta commented:
Folks, it will be really helpful if we can merge this PR quickly. We have new C++ specific functions, that does not exist in java world and native sidecar is the best way to make those discoverable to the coordinator. Our customers are adding C++ functions, and as a work around they are adding dummy java functions that just throws in execution. With the sidecar integration, they can just add the function in worker.

pdabre12 commented:
Updated the approach:
All functionality of the sidecar plugin is now controlled via a configuration file. If the config file is present, the corresponding functionality will be loaded; otherwise, it will not be. This ensures that the build will not fail even if the plugin is included in clusters where sidecar support is disabled.

The type manager loading process remains straightforward and is similar to the function namespace manager loading.

In java clusters and non-sidecar enabled clusters, all worker providers will still be bound as before. However, when sidecar is enabled, `NativeWorkerSessionPropertyProvider ` is now conditionally loaded based on the config. Specifically, when `coordinator-sidecar-enabled` is set to true, the `NativeWorkerSessionPropertyProvider` will not be bound, and all `native-worker` specific properties will be provided via the sidecar.

amitkdutta commented:
@pdabre12 Thanks for iterating on it. Just to clarify a bit more, there is no build failure. In https://github.com/prestodb/presto/issues/24964, we discussed that

1. coordinator always loads the plugin even if `"coordinator-sidecar-enabled": "false"`, this makes coordinator crash
```
025-04-23T12:55:49.520-0700	ERROR	main	com.facebook.presto.server.PrestoServer	System session property provider is already registered for property provider : native-worker
java.lang.IllegalArgumentException: System session property provider is already registered for property provider : native-worker
	at com.facebook.presto.metadata.SessionPropertyManager.loadSessionPropertyProvider(SessionPropertyManager.java:147)
	at com.facebook.presto.metadata.SessionPropertyManager.loadSessionPropertyProviders(SessionPropertyManager.java:154)
	at com.facebook.presto.server.PrestoServer.run(PrestoServer.java:193)
```

Coordinator crashes at runtime, not in build time. We want to control it with  `"coordinator-sidecar-enabled"` property (for example in java clusters even if plugin directory contains this plugin, the property will be set to `false` and won't make coordinator crash at runtime.

pdabre12 commented:
@amitkdutta Yes, my bad, its a runtime issue.
We will control it using `coordinator-sidecar-enabled` as well as using the individual functionalities config files.

To clarify, let's say if users want to use the `NativeFunctionNamespaceManager`, users need to set up the function namespace under `presto-main/etc/function-namespace/`, if not the `NativeFunctionNamespaceManager` won't be used.
With this approach, we can include the `presto-native-sidecar-plugin` in the build directory, but not use its functionalities unless explicitly configured using config files.

```
025-04-23T12:55:49.520-0700	ERROR	main	com.facebook.presto.server.PrestoServer	System session property provider is already registered for property provider : native-worker
java.lang.IllegalArgumentException: System session property provider is already registered for property provider : native-worker
	at com.facebook.presto.metadata.SessionPropertyManager.loadSessionPropertyProvider(SessionPropertyManager.java:147)
	at com.facebook.presto.metadata.SessionPropertyManager.loadSessionPropertyProviders(SessionPropertyManager.java:154)
	at com.facebook.presto.server.PrestoServer.run(PrestoServer.java:193)
```
	
If the `NativeSystemSessionPropertyProvider` which pulls the session properties from the sidecar isn't configured, the coordinator won't attempt to load it and the above coordinator crash would not happen.

rschlussel commented on presto-native-sidecar-plugin/src/test/java/com/facebook/presto/sidecar/TestNativeSidecarPlugin.java:
```diff
@@ -39,7 +39,6 @@
 import static org.testng.Assert.assertFalse;
 import static org.testng.Assert.fail;
 
-@Test(singleThreaded = true)
```
why did you remove this?

pdabre12 replied:
I added this change in. Maybe at the time being I was checking some tests that needed to be executed concurrently, but never pushed those upstream. I don't think there shouldn't be any issue in making this change.

rschlussel reviewed (commented):
approach seems good. At some point when all the functionality is complete/well tested we should change the default. if sidecar is enabled use the sidecar, just to make configuration easier. Because this does make configuration more complex.

rschlussel commented:
actually, this seems like it might become really burdensome for people to configure to use the sidecar vs. just a single config. would it be possible to make the default be to load the sidecar plugin when sidecar is enabled even if there is no e.g. session properties config set.

rschlussel commented:
because previously you just needed to set one config (coordinator-sidecar-enabled).  And now if you want to use the sidecar you need to add new config files for all the different features of the sidecar to say that you want to use the sidecar for it (function manager, session properties manager).  I worry it makes it harder to get started.
